File: codex-rs/core/src/token_metering.rs
use once_cell::sync::Lazy;
use std::fs::{File, OpenOptions};
use std::io::{BufWriter, Write};
use std::sync::Mutex;

use ctor::dtor;
use tiktoken_rs::{ChatCompletionRequestMessage, get_bpe_from_model, num_tokens_from_messages};

/// Environment variable that enables token logging when set to a file path.
pub const TOKEN_LOG_ENV_VAR: &str = "CODEX_TOKEN_LOG";

static TOKEN_LOGGER: Lazy<Option<Mutex<BufWriter<File>>>> = Lazy::new(|| {
    let path = std::env::var(TOKEN_LOG_ENV_VAR).ok()?;
    let file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(path)
        .ok()?;
    Some(Mutex::new(BufWriter::new(file)))
});

#[dtor]
fn flush_on_exit() {
    flush_log();
}

/// Append a token usage record to the CSV log if enabled.
pub fn record_usage(model: &str, prompt_tokens: u64, completion_tokens: u64) {
    if let Some(ref mutex) = *TOKEN_LOGGER {
        if let Ok(mut writer) = mutex.lock() {
            let _ = writeln!(writer, "{model},{prompt_tokens},{completion_tokens}");
        }
    }
}

/// Flush the log to disk if logging is enabled.
pub fn flush_log() {
    if let Some(ref mutex) = *TOKEN_LOGGER {
        if let Ok(mut writer) = mutex.lock() {
            let _ = writer.flush();
        }
    }
}

/// Count the number of tokens in chat messages using tiktoken.
pub fn count_prompt_tokens(
    model: &str,
    messages: &[ChatCompletionRequestMessage],
) -> Option<usize> {
    num_tokens_from_messages(model, messages).ok()
}

/// Count tokens for plain text using tiktoken.
pub fn count_text_tokens(model: &str, text: &str) -> Option<usize> {
    let bpe = get_bpe_from_model(model).ok()?;
    Some(bpe.encode_with_special_tokens(text).len())
}

File: codex-cli/src/utils/token-metering.ts
export const TOKEN_LOG_ENV_VAR = "CODEX_TOKEN_LOG";

import { appendFileSync } from "fs";
import type { OpenAI } from "openai";

export function recordTokenUsage(
  model: string,
  promptTokens: number,
  completionTokens: number,
): void {
  const path = process.env[TOKEN_LOG_ENV_VAR];
  if (!path) return;
  try {
    appendFileSync(path, `${model},${promptTokens},${completionTokens}\n`);
  } catch {
    // ignore write errors
  }
}

export function countPromptTokens(
  messages: Array<OpenAI.Chat.Completions.ChatCompletionMessageParam>,
): number {
  let chars = 0;
  for (const m of messages) {
    if (typeof m.content === "string") {
      chars += m.content.length;
    }
    const fc = (m as unknown as { function_call?: { name: string; arguments: string } }).function_call;
    if (fc) {
      chars += fc.name.length + fc.arguments.length;
    }
  }
  return Math.ceil(chars / 4);
}

export function countTextTokens(text: string): number {
  return Math.ceil(text.length / 4);
}
